src/deploy.ts

```diff
@@ -190,14 +203,128 @@ class Deployer {
+      if (isGit) {
+        const remotes = (await promisify(exec)("git remote -v", {cwd: this.deployOptions.config.root})).stdout
+          .split("\n")
+          .filter((d) => d)
+          .map((d) => d.split(/\s/g));
+        const gitHub = remotes.find(([, url]) => url.startsWith("https://github.com/"));
+        if (gitHub) {
+          const repoName = formatGitUrl(gitHub[1]);
+          const repositories = (await this.apiClient.getGitHubRepositories())?.repositories;
+          const authedRepo = repositories?.find(({url}) => formatGitUrl(url) === repoName);
+          if (authedRepo) {
+            // Set branch to current branch
+            const branch = (
+              await promisify(exec)("git rev-parse --abbrev-ref HEAD", {cwd: this.deployOptions.config.root})
+            ).stdout;
+            await this.apiClient.postProjectEnvironment(deployTarget.project.id, {
+              source: {
+                provider: authedRepo.provider,
+                provider_id: authedRepo.provider_id,
+                url: authedRepo.url,
+                branch
+              }
+            });
+            return true;
+          } else {
+            // repo not auth’ed; link to auth page and poll for auth
+            // TODO: link to internal page that bookends the flow and handles the no-oauth-token case more gracefully
+            this.effects.clack.log.info(
+              `Authorize Observable to access the ${bold(repoName)} repository: ${link(
+                "https://github.com/apps/observable-data-apps-dev/installations/select_target"
+              )}`
+            );
+            const spinner = this.effects.clack.spinner();
+            spinner.start("Waiting for repository to be authorized");
+            const pollExpiration = Date.now() + DEPLOY_POLL_MAX_MS;
+            while (true) {
+              if (Date.now() > pollExpiration) {
+                spinner.stop("Waiting for repository to be authorized timed out.");
+                throw new CliError("Repository authorization failed");
+              }
+              const repositories = (await this.apiClient.getGitHubRepositories())?.repositories;
+              const authedRepo = repositories?.find(({url}) => formatGitUrl(url) === repoName);
+              if (authedRepo) {
+                spinner.stop("Repository authorized.");
+                await this.apiClient.postProjectEnvironment(deployTarget.project.id, {
+                  source: {
+                    provider: authedRepo.provider,
+                    provider_id: authedRepo.provider_id,
+                    url: authedRepo.url,
+                    branch: null // TODO detect branch
+                  }
+                });
+                return true;
+              }
+              await new Promise((resolve) => setTimeout(resolve, 2000));
+            }
+          }
+        } else {
+          this.effects.clack.log.error("No GitHub remote found; cannot enable continuous deployment.");
```

Comment:
these should exit with a CliError rather than just logging a message and continuing on with a local deploy

Reply:
done